vcap_utils/vcap_utils/backends/base_openvino.py

```diff
@@ -35,34 +35,112 @@ def __init__(self, model_xml: bytes,
+                device_name=device_name,
+                num_requests=num_requests)
+        except RuntimeError as e:
+            if device_name == "CPU":
+                # It's unknown why this would happen when loading
+                # onto CPU, so we re-raise the error.
+                raise e
+
+            # This error happens when trying to load onto HDDL or MYRIAD, but
+            # somehow the device fails to work. In these cases, we try again
+            # but load onto 'CPU', which is a safe choice.
+            msg = f"Failed to load {self.__class__} onto device " \
+                  f"{device_name}. Error: '{e}'. " \
+                  f"Trying again with device_name='CPU'"
+            logging.critical(msg)
+            self.__init__(
+                model_xml=model_xml,
+                weights_bin=weights_bin,
+                device_name="CPU",
+                ie_core=ie_core)
+            return
 
         # Pull out a couple useful constants
-        self.InferRequestStatusCode = StatusCode
         self.input_blob_names: List[str] = list(self.net.inputs.keys())
         self.output_blob_names: List[str] = list(self.net.outputs.keys())
 
+        # For running threaded requests to the network
+        self._get_free_request_lock = RLock()
+        self._request_free_events: List[Event] = [
+            Event() for _ in self.exec_net.requests]
+        for request_free in self._request_free_events:
+            request_free.set()
+
+        # For keeping track of the workload for this backend
+        self._total_requests: int = len(self.exec_net.requests)
+        self._num_ongoing_requests: int = 0
+        self._num_ongoing_requests_lock: RLock = RLock()
+
+    @property
+    def workload(self) -> float:
+        """Returns the percent saturation of this backend. The backend is
+        at max 'efficiency' once the number of ongoing requests is equal to
+        or over number of exec_net.requests
+        """
+        return self._num_ongoing_requests / self._total_requests
+
+    def send_to_batch(self, input_data: OV_INPUT_TYPE) -> Queue:
+        """Efficiently send the input to be inferenced by the network
+        :param input_data: Input to the network
+        :returns: A queue that will yield 1 result, the output from the network
+        """
+        out_queue = Queue(maxsize=1)
+
+        with self._get_free_request_lock:
+            # Try to get at least one request
+            request_id = self.exec_net.get_idle_request_id()
+            if request_id < 0:
+                # Since there was no free request, wait for one
+                self.exec_net.wait(num_requests=1)
+                request_id = self.exec_net.get_idle_request_id()
+                if request_id < 0:
+                    raise RuntimeError("Invalid request ID!")
+
+            request = self.exec_net.requests[request_id]
+            request_free = self._request_free_events[request_id]
+
+            def on_result(*args):
```

Comment:
This callback is assigned to the requests completion callback, at which point it promptly feeds the networks results into the output_queue, and calls request_free.set() so that the request can be used by another thread.

Reply:
Could you create a new event object in send_to_batch for each call instead of maintaining the list of event objects (`_request_free_events`)? It seems like these events' lifecycles are just within this method, though I may be missing a usage.

Reply:
Actually, the Event lifecycle is outside of this method. There's 1 event for each InferRequest, and I use them to make sure that there aren't any callbacks currently being run for a past request, which could cause a race condition where set_completion_callback is replaced before the last request has gotten a chance to call it. 

IMO this is a problem with OpenVINO...

Reply:
I've added a comment explaining what _request_free_events is for